### PR TITLE
Adds basic keybinding for focusing on an element in outline pane

### LINKED ISSUE #90732: Outline view is missing the keyboard command to just move the editor to the current a symbol
1 In Outline View, mouse-click on a symbol e.g,.
image
2 See the editor moves to that symbol
3 Press Down Arrow to move to another symbol.
4 Try to find a keypress that just moves the editor to that symbol.

Expected: success
Observed: failure, here. There's none I can find by experiment or reading docs. The nearest is Enter, but that also switches focus away.

Check the Command Palette. Neither this or any other commands to operate Outline View are present.

V1.42.1 on Windows.

### PR DESCRIPTION
**Pull request by lambainsaan**
This PR fixes #90732.

You can test out the change with the new key binding added under the name `list.focus` for the same feature.

### CODE REVIEW

**Comment by jrieken** [comment]
Outline changes look legit - to @bpasero for changes on the list commands

**Review by bpasero — CHANGES_REQUESTED** [review]
This is very hard to understand until you realize that you added a `list.focus` command that is pretty much a copy of `list.select` with just a few modifications. Can you extract all the code of both commands into 1 reusable method and make the things that are actually different adjustable from the outside?

I also see that you changed from:

`const fakeKeyboardEvent = getSelectionKeyboardEvent('keydown', false);`

to

`const fakeKeyboardEvent = getSelectionKeyboardEvent('keydown', true);`

for the existing `list.select`. What is the reasoning?

**Comment by lambainsaan** [comment]
> This is very hard to understand until you realize that you added a `list.focus` command that is pretty much a copy of `list.select` with just a few modifications. Can you extract all the code of both commands into 1 reusable method and make the things that are actually different adjustable from the outside?

Agreed, I was thinking on the same lines, but I refrained from doing it. I will update the PR.

> 
> I also see that you changed from:
> 
> `const fakeKeyboardEvent = getSelectionKeyboardEvent('keydown', false);`
> 
> to
> 
> `const fakeKeyboardEvent = getSelectionKeyboardEvent('keydown', true);`
> 
> for the existing `list.select`. What is the reasoning?

Oh yes, looking back at it, it does not make sense to me. It should not have changed. I will switch to the older version for this, and flip the usage.

**Comment by lambainsaan** [comment]
Left it hanging for quite sometime, @bpasero could you please take a look at it again? :)

File: src/vs/workbench/browser/actions/listCommands.ts [hunk 1]
```diff
@@ -494,6 +494,41 @@ function listFocusLast(accessor: ServicesAccessor, options?: { fromFocused: bool
...
+function focusElement(accessor: ServicesAccessor, retainCurrentFocus: boolean): void {
+	const focused = accessor.get(IListService).lastFocusedList;
+
+	// List
+	if (focused instanceof List || focused instanceof PagedList) {
+		const list = focused;
+		list.setSelection(list.getFocus());
+	}
+
+	// ObjectTree
```
Comment (bpasero) [inline src/vs/workbench/browser/actions/listCommands.ts:507]:
```suggestion
	// Trees
```

File: src/vs/workbench/browser/actions/listCommands.ts [hunk 2]
```diff
@@ -494,6 +494,41 @@ function listFocusLast(accessor: ServicesAccessor, options?: { fromFocused: bool
...
 }
 
+
+function focusElement(accessor: ServicesAccessor, retainCurrentFocus: boolean): void {
+	const focused = accessor.get(IListService).lastFocusedList;
+
+	// List
+	if (focused instanceof List || focused instanceof PagedList) {
+		const list = focused;
+		list.setSelection(list.getFocus());
```
Comment (bpasero) [inline src/vs/workbench/browser/actions/listCommands.ts:504]: This used to be:

```ts
const list = focused;
list.setSelection(list.getFocus(), fakeKeyboardEvent);
list.open(list.getFocus(), fakeKeyboardEvent);
```

Why this change?

**Review by bpasero — CHANGES_REQUESTED** [review]

File: src/vs/workbench/browser/actions/listCommands.ts (hunk 2 above)
Comment (lambainsaan) [reply src/vs/workbench/browser/actions/listCommands.ts]: Ah damn it, lost this change while rebasing. Updated the PR.

File: src/vs/workbench/contrib/outline/browser/outlinePane.ts [hunk 3]
```diff
@@ -571,7 +571,8 @@ export class OutlinePane extends ViewPane {
 			// todo@Joh
 			if (e.browserEvent) {
 				if (e.browserEvent.type === 'keydown') {
-					focus = true;
+					const selectionEvent = e.browserEvent as SelectionKeyboardEvent;
+					focus = typeof selectionEvent.preserveFocus === 'boolean' ? selectionEvent.preserveFocus : true;
```
Comment (bpasero) [inline src/vs/workbench/contrib/outline/browser/outlinePane.ts:575]:
```suggestion
					focus = typeof selectionEvent.preserveFocus === 'boolean' ? !selectionEvent.preserveFocus : true;
```

The boolean logic is flawed here I think, can you please test this.

**Review by bpasero — CHANGES_REQUESTED** [review]

**Comment by bpasero** [comment]
Also adding @joaomoreno to comment on the fact that we are about to introduce a new list/tree command `list.focus` that only differs from `list.select` in that it will retain focus.

I wonder if the name should actually be `list.selectKeepFocus` or so, because I think no-one really understands the difference between `focus` and `select`?

**Comment by joaomoreno** [comment]
As you move up/down in the outline (and any tree/list) you're moving the **focus**. If you press `Enter`, you're **selecting** the focused element. When things get selected, editors open. Also, it just so happens that we've made a decision to always move DOM focus away from the list/tree and into the editor when this happens. So, we need to add a new keybinding which does the same but preserves the DOM focus.

Calling it `list.focus` is indeed a bad name. I would call it `list.selectAndPreserveFocus` to match `list.select`.

---

@bpasero The concept exists everywhere. Here's windows:

image

All those folders are selected, but only `recipes` is focused.

**Comment by bpasero** [comment]
Yeah, I like `list.selectAndPreserveFocus`. @lambainsaan can you make these changes?

File: src/vs/workbench/contrib/outline/browser/outlinePane.ts (hunk 3 above)
Comment (lambainsaan) [reply src/vs/workbench/contrib/outline/browser/outlinePane.ts]: We don't really need this check, the old code was using it, and I forgot to remove this code while rebasing.

**Comment by lambainsaan** [comment]
Sure thing, @bpasero , I have made the changes.

**Comment by bpasero** [comment]
Thanks!
